crates/sui-core/src/safe_client.rs

```diff
@@ -398,13 +401,47 @@ where
         Ok(transaction_info)
     }
 
+    fn verify_authenticated_checkpoint(&self, checkpoint: &AuthenticatedCheckpoint) -> SuiResult {
+        match checkpoint {
+            AuthenticatedCheckpoint::None => Ok(()),
+            AuthenticatedCheckpoint::Signed(s) => s.verify(),
+            AuthenticatedCheckpoint::Certified(c) => c.verify(&self.committee),
+        }
+    }
+
     pub async fn handle_checkpoint(
         &self,
         request: CheckpointRequest,
     ) -> Result<CheckpointResponse, SuiError> {
-        // SECURITY TODO: Implement all checks!
+        let req_type = request.request_type.clone();
 
-        self.authority_client.handle_checkpoint(request).await
+        let resp = self.authority_client.handle_checkpoint(request).await?;
+
+        match req_type {
+            CheckpointRequestType::LatestCheckpointProposal => {
+                if let AuthorityCheckpointInfo::Proposal { current, previous } = &resp.info {
+                    if let Some(current) = current {
+                        current.verify()?;
+                    }
+                    self.verify_authenticated_checkpoint(previous)?;
+                    Ok(resp)
+                } else {
+                    Err(SuiError::ByzantineAuthoritySuspicion {
+                        authority: self.address,
+                    })
+                }
+            }
+            CheckpointRequestType::PastCheckpoint(_) => {
```

Comment:
Should we check that the checkpoint returned is the correct sequence as requested?

Reply:
might as well!